Feature/add cliente implementation

HenriqueMitsuo opened the pull request.

HenriqueMitsuo committed:
chore: update package-lock

HenriqueMitsuo committed:
chore: update eslint unused rule for typescript

HenriqueMitsuo committed:
feat(domain): add base ValueObject

HenriqueMitsuo committed:
feat(AssertionConcern): add new validation methods
- assertArgumentIsValidEmail: validação de email
- assertArgumentIsValidCpf: validação de cpf
- assertArgumentIsObject: verifica se é um objeto
- assertObjectEquality: compara dois objetos

HenriqueMitsuo committed:
feat(domain): add Cliente entity and its valueObjects

HenriqueMitsuo committed:
feat(domain): add Cliente ports definitions

HenriqueMitsuo committed:
chore: update to a named export

HenriqueMitsuo committed:
feat(adapters): add Cliente mongo repository

HenriqueMitsuo committed:
feat(api): add Cliente feature

CarolPinheiro commented on src/adapter/driver/api/controllers/cliente/index.ts:
```diff
@@ -0,0 +1,52 @@
+        try {
+            const email = req.query.email as string;
+
+            if (!email) return res.status(401).json();
```
**suggestion:** Como o e-mail está vindo via query string, e não é algo obrigatório ou tão claro pra rota, não seria interessante retornarmos aqui uma mensagem para o usuário explicando o problema?

CarolPinheiro commented on src/adapter/driver/api/controllers/cliente/index.ts:
```diff
@@ -0,0 +1,52 @@
+
+    public async getByCpf(req: Request, res: Response): Promise<Response> {
+        try {
+            const cpf = req.query.cpf as string;
+
+            if (!cpf) return res.status(401).json();
```
**suggestion:** Mesma sugestão para retornarmos uma mensagem mais clara. 

E uma coisa que fiquei refletindo (não sei se aplicaria tbm pro e-mail), mas não faria sentido usarmos o CPF como pathParameter ao invés de query param? Considerando que estamos buscando um usuário único.

CarolPinheiro commented on src/core/application/useCases/user/index.ts:
```diff
@@ -1,16 +1,19 @@
 
-export class SystemUserUseCase implements UserUseCase {
-    private readonly userRepository: UserRepository;
+    public create(cliente: Cliente): Promise<Cliente> {
+        return this.clienteRepository.create(cliente);
+    }
```
**suggestion:** Questão que eu pensei aqui, não sei como vai ser a questão pra e-mail, mas CPF não poderíamos deixar repetir.

Acho que seria interessante antes de criar fazermos uma busca com o CPF e talvez o e-mail, pra garantir que a pessoa pode criar essa conta.

CarolPinheiro commented on src/core/domain/base/assertionConcern.ts:
```diff
@@ -38,4 +38,56 @@ export class AssertionConcern {
+        message: string,
+    ): void {
+        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+        console.log(email);
+        console.log(emailRegex.test(email));
```
**nitpick:** Acho que podemos remover esses consoles.

CarolPinheiro commented on src/adapter/driver/api/types/cliente-request.ts:
```diff
@@ -0,0 +1,5 @@
+export interface CreateClienteRequestBody {
```
**nitpick:** Aqui acho que seria interessante o arquivo seguir o padrão de nomenclatura de `clienteRequest.interface.ts`

E não seria interessante mantermos o padrão de nomenclatura como ICreateClienteRequestBody?

HenriqueMitsuo replied:
ali foi por causa do conflito de nomes, não acho interessante esse padrão de nomeclatura

CarolPinheiro replied:
Falando pelo menos do nome do arquivo, eu colocaria ele no mínimo em camelCase, pra seguir o resto do projeto.

CarolPinheiro commented on src/core/application/useCases/user/cliente.useCase.ts:
```diff
@@ -0,0 +1,7 @@
+import { Cliente } from "@domain/entities/cliente";
+
+export interface IClienteUseCase {
```
**suggestion:** Aqui pra seguir a padronização que combinamos, seria interessante termos o nome do arquivo como cliente.interface.ts

CarolPinheiro reviewed (changes requested):
Deixei algumas pequenas sugestões de alteração

CarolPinheiro commented on src/core/application/useCases/user/cliente.useCase.ts:
```diff
@@ -0,0 +1,7 @@
+import { Cliente } from "@domain/entities/cliente";
```
**nitpick:** Micro detalhe, só precisamos alterar o nome dessa pasta para cliente.

HenriqueMitsuo committed:
chore: rename file

HenriqueMitsuo committed:
feat(ClienteController): update get routes
- update to param usage
- add missing identifier error message

HenriqueMitsuo committed:
chore: remove console.log

HenriqueMitsuo committed:
feat(ClienteUseCase): update create
- add check for duplicate rule

HenriqueMitsuo committed:
chore: rename file

CarolPinheiro commented on src/core/application/useCases/user/index.ts:
```diff
@@ -1,16 +1,26 @@
-import { UserRepository } from "@domain/repositories/userRepository";
-import { UserUseCase } from "./types";
-import { User } from "@domain/entities/user";
+import { Cliente } from "@domain/entities/cliente";
```
**nitpick:** Última coisa, só faltou mudar o nome da pasta pra cliente.

karineriquena replied:
o que é nitpick? curiosidade só rs

CarolPinheiro replied:
@karineriquena nitpick é tipo, um micro detalhe kkk como mudar nome de pasta, etc

CarolPinheiro reviewed (changes requested):
@HenriqueMitsuo hey, uma coisa que eu acabei notando só agora (erro meu), é que acabou faltando os testes. Acho que podemos seguir a estrutura que eu fiz em Produtos, ou se tiver alguma outra ideia seria maneiro termos isso.

karineriquena commented on src/adapter/driven/infra/mongo/models/Cliente.ts:
```diff
@@ -0,0 +1,27 @@
+        },
+        cpf: {
+            type: String,
+            required: false,
```
Só fiquei na duvida pois no tech challenge fala sobre se identificar com CPF e se cadastrar com nome + email... Então como a identificação é pelo CPF, imagino que seja required = true:

"Os clientes são apresentados a uma interface de seleção na qual podem optar por se identificarem via CPF, se cadastrarem com nome, e-mail ou não se identificar"

O que eu imaginei: Uma tela com as duas opções "Se identificar" e "Pular e selecionar combo", sendo a opção para identificar por CPF, caso o CPF exista, traz o nome e -mail preenchidos, caso não, a pessoa preenche o nome e email.

Outra opção possível que eu imaginei seria a identificação ocorrer por "CPF" ou "E-MAIL" que são as opções que não se repetiriam numa base de dados

HenriqueMitsuo replied:
acho que faz sentido, tinha interpretado que era um cpf opcional

karineriquena commented on src/adapter/driver/api/routes/clienteRouter.ts:
```diff
@@ -0,0 +1,18 @@
+
+    clienteRouter.post("/", async (req, res) =>
+        clienteController.post(req, res),
+    );
```
sugestão (pode ignorar se quiser):
```suggestion
    );
   
```

karineriquena commented on src/adapter/driver/api/routes/clienteRouter.ts:
```diff
@@ -0,0 +1,18 @@
+    );
+    clienteRouter.get("/by-email/:email", async (req, res) =>
+        clienteController.getByEmail(req, res),
+    );
```
```suggestion
    );
    
```

karineriquena commented on src/adapter/driver/api/routes/index.ts:
```diff
@@ -1,11 +1,11 @@
     const serverRouter = Router();
 
-    serverRouter.use("/users", getUserRouter());
+    serverRouter.use("/cliente", makeClienteRouter());
```
a rota de produto está no plural e a de cliente no singular, deveríamos padronizar?

HenriqueMitsuo replied:
vou trocar a de "produtos", o correto é singular para rest

karineriquena commented on src/adapter/driver/api/types/clienteRequest.ts:
```diff
@@ -0,0 +1,5 @@
+export interface CreateClienteRequestBody {
```
vamos seguir com o ".interface" no nome do arquivo?

clientRequest.interface.ts

HenriqueMitsuo replied:
Precisaria mesmo? não entendi muito bem a utilidade de colocar interface no nome do arquivo, a impressão é que isso resolve um problema muito simples talvez no inicio de um projeto, quando não se tem contexto da estrutura e/ou typescript, isso vem de algum guia? do que já vi acho que era alguma convensão do angular
- o arquivo já está em uma pasta types

karineriquena replied:
é que a gente colocou ".interface" em outros... e sim acho que é do angular colocar .interface, .module, .service etc... mas vou fechar aqui então

karineriquena commented on src/core/application/useCases/user/index.ts:
```diff
@@ -1,16 +1,26 @@
-export class SystemUserUseCase implements UserUseCase {
-    private readonly userRepository: UserRepository;
+    public async create(cliente: Cliente): Promise<Cliente> {
+        const jaExiste = await this.clienteRepository.verificarDuplicado({
```
```suggestion
        const alreadyExists= await this.clienteRepository.verificarDuplicado({
```

karineriquena commented on src/core/domain/base/valueObject.ts:
```diff
@@ -1,5 +1,25 @@
+            return false;
+        }
+
+        if (valueObject.props === undefined) {
```
precisa do "=== undefined"?  ou pode ser:

```typescript
if (valueObject.props) {
```

HenriqueMitsuo committed:
chore: add additional path alias

HenriqueMitsuo committed:
feat(api): add request logger

HenriqueMitsuo committed:
chore: rename infra adpater to be more specific

HenriqueMitsuo committed:
chore: remove unecessary prefix

HenriqueMitsuo committed:
chore: rename files

HenriqueMitsuo committed:
feat(Cliente): add remaining fields validation

HenriqueMitsuo committed:
refactor(ClienteRepository):
- make email and cpf unique on mongo implementation
- update return to match interface
- renamed verificarDuplicado to checkDuplicate

HenriqueMitsuo committed:
feat(api): add StatusCode constant util

HenriqueMitsuo committed:
feat(api): add clienteMapper

HenriqueMitsuo committed:
refactor(ClienteController):
- update controller format
- add statusCode constant usage
- update usaCase usage

HenriqueMitsuo committed:
chore: organize imports

HenriqueMitsuo committed:
chore: rename route to follow rest standard

HenriqueMitsuo committed:
chore: update interface usage

HenriqueMitsuo committed:
test: add cliente related tests

daniel-eiro commented:
Boa Henrique, dei uma olhada nos códigos e ficou muito bom. Eu ia comentar sobre os retornos dos status code (tinha um 401 ali que eu não estava entendendo muito bem), mas vi que você alterou para um arquivo padronizado, muito bacana, parabéns!
